### PR TITLE
Core: Support manager build with webpack5

### PR DESCRIPTION
**Pull request by shilman**
Issue: #14044 

**NOTE:** I released `6.3.0-alpha.37` / `38` / `39` off this feature branch for testing purposes!

## What I did

- [x] Extract manager-webpack4 from core-server
- [x] Add manager-webpack5 that uses webpack5-compatible plugins
- [x] Other minor refactors
- [x] Fix warning ("expected 4, received 4")
- [x] `import * as webpackReal` for addons https://github.com/storybookjs/storybook/pull/15016
- [x] Add warning when you don't have `manager-webpack5`, link to `MIGRATION.md`

Self-merging @tmeasday @ndelangen @wKich @yannbf 

## How to test

- [x] CI passes (ish)
- [x] Released on alpha & tested in sample app

```sh
yarn add @storybook/builder-webpack5 @storybook/manager-webpack5 webpack --dev
```

Then update `.storybook/main.js`:

```js
module.exports = {
  core: {
    builder: 'webpack5'
  }
}
```

### CODE REVIEW

File: lib/builder-webpack4/src/index.ts [hunk 1]
```diff
@@ -1,20 +1,31 @@
...
 let compilation: ReturnType<typeof webpackDevMiddleware>;
 let reject: (reason?: any) => void;
 
 type WebpackBuilder = Builder<Configuration, Stats>;
 
-const webpack = (webpackReal as any) as typeof webpack4;
-
-const checkWebpackVersion4 = (webpackInstance: { version?: string }) =>
-  checkWebpackVersion(webpackInstance, '4.x', 'builder-webpack4');
+// const webpack = (webpackReal as any) as typeof webpackType;
```
Comment (yannbf) [inline lib/builder-webpack4/src/index.ts:19]: Was this left here intentionally?

File: lib/core-server/src/core-presets.test.ts [hunk 2]
```diff
@@ -24,26 +24,28 @@ jest.setTimeout(10000);
...
 const skipStoriesJsonPreset = [{ features: { buildStoriesJson: false } }];
 
-jest.mock('@storybook/builder-webpack5', () => {
-  const actualBuilder = jest.requireActual('@storybook/builder-webpack5');
-  // MUTATION! we couldn't mock webpack5, so we added a level of indirection instead
-  actualBuilder.executor.get = jest.fn();
-  actualBuilder.overridePresets = [...actualBuilder.overridePresets, skipStoriesJsonPreset];
-  return actualBuilder;
-});
+// jest.mock('@storybook/builder-webpack5', () => {
```
Comment (yannbf) [inline lib/core-server/src/core-presets.test.ts:27]: Should this be commented out?

File: lib/core-server/src/core-presets.test.ts [hunk 3]
```diff
@@ -170,8 +172,8 @@ describe.each([
   });
 });
 
-const progressPlugin = (config) =>
-  config.plugins.find((p) => p.constructor.name === 'ProgressPlugin');
+// const progressPlugin = (config) =>
```
Comment (yannbf) [inline lib/core-server/src/core-presets.test.ts:175]: intentional comments?

File: lib/core-server/src/core-presets.test.ts [hunk 4]
```diff
@@ -181,18 +183,18 @@ describe('dev cli flags', () => {
...
-    expect(progressPlugin(config)).toBeTruthy();
-  });
-
-  it('--quiet', async () => {
-    const options = { ...cliOptions, quiet: true };
-    await buildDevStandalone(options);
-    const config = getConfig(previewExecutor.get, 'preview');
-    expect(progressPlugin(config)).toBeFalsy();
-  });
+  // it('baseline', async () => {
```
Comment (yannbf) [inline lib/core-server/src/core-presets.test.ts:186]: intentional comments? If they are to be resolved later maybe `.skip` + a comment with a reason?

File: lib/core-server/src/utils/get-manager-builder.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  const main = path.resolve(configDir, 'main');
+  const mainFile = getInterpretedFile(main);
+  const { core } = mainFile ? serverRequire(mainFile) : { core: null };
+
+  const builderPackage =
+    core?.builder === 'webpack5' ? '@storybook/manager-webpack5' : '@storybook/manager-webpack4';
+
+  const managerBuilder = await import(builderPackage);
+  return managerBuilder;
+}
```
Comment (yannbf) [inline lib/core-server/src/utils/get-manager-builder.ts:14]: What about something like this? it's a bit more scalable if there are going to be more than two builders:

```suggestion
import path from 'path';
import { getInterpretedFile, serverRequire, Options } from '@storybook/core-common';

const builderVersions = {
  webpack4: '@storybook/manager-webpack4',
  webpack5: '@storybook/manager-webpack5',
}

export async function getManagerBuilder(configDir: Options['configDir']) {
  const main = path.resolve(configDir, 'main');
  const mainFile = getInterpretedFile(main);

  if(!mainFile) {
    // use webpack4 by default
    return import(builderVersions.webpack4);
  }
  
  const { core } = serverRequire(mainFile);

  const builderPackage = builderVersions[core.builder];

  const managerBuilder = await import(builderPackage);
  return managerBuilder;
}
```

File: lib/manager-webpack5/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+        process.exitCode = 1;
+        fail(error || stats);
+      } else {
+        logger.trace({ message: '=> Manager built', time: process.hrtime(startTime) });
+        if (stats && stats.hasWarnings()) {
+          stats.toJson({ warnings: true }).warnings.forEach((e) => logger.warn(e.message));
+        }
+
+        // const statsData = stats.toJson(
```
Comment (yannbf) [inline lib/manager-webpack5/src/index.ts:159]: intentional comment?

File: lib/manager-webpack5/src/manager-webpack.config.ts [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        template,
+      }) as any) as WebpackPluginInstance,
+      (new CaseSensitivePathsPlugin() as any) as WebpackPluginInstance,
+      (new Dotenv({ silent: true }) as any) as WebpackPluginInstance,
+      // graphql sources check process variable
+      new DefinePlugin({
+        'process.env': stringifyEnvs(envs),
+        NODE_ENV: JSON.stringify(envs.NODE_ENV),
+      }) as WebpackPluginInstance,
+      // isProd &&
```
Comment (yannbf) [inline lib/manager-webpack5/src/manager-webpack.config.ts:113]: intentional comment?

File: lib/core-server/src/utils/get-manager-builder.ts (hunk 5 above)
Comment (shilman) [reply lib/core-server/src/utils/get-manager-builder.ts:14]: Added clarifying comment in the code!

**Pull request merged**